core/services/blockheaderfeeder/block_header_feeder.go

```diff
@@ -0,0 +1,247 @@
+		lggr.Info("Stored earliest block number")
+		return nil
+	}
+
+	blocks, err := blockhashstore.DecreasingBlockRange(earliestStoredBlockNumber.Sub(earliestStoredBlockNumber, big.NewInt(1)), minBlockNumber)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(blocks); i += int(f.storeBlockhashesBatchSize) {
+		j := i + int(f.storeBlockhashesBatchSize)
+		if j > len(blocks) {
+			j = len(blocks)
+		}
+		blockRange := blocks[i:j]
+		blockHeaders, err := f.getRlpHeadersBatch(ctx, blockRange)
+		if err != nil {
+			return errors.Wrap(err, "fetching block headers")
+		}
+
+		lggr.Debugw("storing block headers", "blockRange", blockRange)
+		err = f.batchBHS.StoreVerifyHeader(ctx, blockRange, blockHeaders)
+		if err != nil {
+			return errors.Wrap(err, "storing block headers")
+		}
+	}
+
+	if f.lastRunBlock != 0 {
+		// Prune stored, anything older than fromBlock can be discarded
+		for block := f.lastRunBlock - uint64(f.lookbackBlocks); block < uint64(fromBlock); block++ {
+			if _, ok := f.stored[block]; ok {
+				delete(f.stored, block)
+				f.lggr.Debugw("Pruned block from stored cache",
+					"block", block, "latestBlock", latestBlock)
+			}
+		}
+	}
+	f.lastRunBlock = latestBlock
+	return nil
+}
+
+func (f *BlockHeaderFeeder) findLowestBlockNumberWithoutBlockhash(ctx context.Context, lggr logger.Logger, blockToRequests map[uint64]map[string]struct{}) *big.Int {
+	var min *big.Int
+	for block, unfulfilledReqs := range blockToRequests {
+		if len(unfulfilledReqs) == 0 {
+			continue
+		}
+		if _, ok := f.stored[block]; ok {
+			// Already stored
+			continue
+		}
+		stored, err := f.bhs.IsStored(ctx, block)
+		if err != nil {
+			lggr.Warnw("Failed to check if block is already stored",
+				"error", err,
+				"block", block)
+		} else if stored {
+			lggr.Infow("Blockhash already stored",
+				"block", block, "unfulfilledReqIDs", blockhashstore.LimitReqIDs(unfulfilledReqs))
+			f.stored[block] = struct{}{}
+			continue
+		}
+		blockNumber := big.NewInt(0).SetUint64(block)
+		if min == nil || min.Cmp(blockNumber) >= 0 {
+			min = blockNumber
+		}
+	}
+	return min
+}
+
+func (f *BlockHeaderFeeder) findEarliestBlockNumberWithBlockhash(ctx context.Context, lggr logger.Logger, startBlock, toBlock uint64) (*big.Int, error) {
+	from := startBlock
+	for i := startBlock; i < toBlock; i += uint64(f.getBlockhashesBatchSize) {
+		j := from + uint64(f.getBlockhashesBatchSize)
+		if j > toBlock {
+			j = toBlock
+		}
+
+		lggr.Debug(fmt.Sprintf("Looking for earliest block number with blockhash %v thru %v", i, j))
+
+		blockNumber := i
+		var blocks []*big.Int
+		for blockNumber < j {
+			blocks = append(blocks, big.NewInt(0).SetUint64(blockNumber))
+			blockNumber++
+		}
+
+		blockhashes, err := f.batchBHS.GetBlockhashes(ctx, blocks)
+		if err != nil {
+			return nil, errors.Wrap(err, "fetching blockhashes")
+		}
+
+		for idx, bh := range blockhashes {
+			if !bytes.Equal(bh[:], zeroHash[:]) {
+				earliestBlockNumber := i + uint64(idx)
```

Comment:
Just confirming the index arithmetic here, this will be `i + 0` since `idx` starts at 0, I think that's fine right?

Reply:
yeah should be fine. `startBlock` is inclusive and `i` is also inclusive